lib/usermanager.js

```diff
@@ -340,36 +340,36 @@ exports = module.exports = {
   },
 
   /**
-    * handle a users unshared courses on delete
-  */
+   * Handle a users courses on delete
+   */
   handleUserCoursesOnDelete: function(userId, courseOption, done) {
     var pluginManager = pluginmanager.getManager();
     var plugins = pluginManager.getPlugins();
     var CourseContent = require(plugins.content.course.fullPath);
-
     var courseContent = new CourseContent();
-    courseContent.retrieve({ createdBy: userId, _isShared: false }, {}, function (err, results) {
-
-      switch (courseOption) {
-        case 'transfer':
-          var currentUser = this.getCurrentUser();
-          async.each(results, function(model, cb) {
-            courseContent.update({ _id: model._id }, { createdBy: currentUser._id }, cb);
-          }, done);
-          break;
-        case 'delete':
-          async.each(results, function(model, cb) {
-            courseContent.destroy({ _id: model._id }, false, cb);
-          }, done);
-          break;
-        case 'share':
-          async.each(results, function(model, cb) {
-            courseContent.update({ _id: model._id }, { _isShared: true }, cb);
-          }, done);
-          break;
-        default:
-          done(null);
-      }
+    if (courseOption === null) {
+      return done(null);
+    }
+    courseContent.retrieve({ createdBy: userId }, {}, function (err, results) {
+      var currentUser = this.getCurrentUser();
+      async.eachSeries(results, function(model, cb) {
+        // Transfer ownership of all courses to current user
+        courseContent.update({ _id: model._id }, { createdBy: currentUser._id }, function (err, course) {
+          if (!course._isShared && course._shareWithUsers.length === 0) {
```

Comment:
Is it possible to reverse the logic here to cut out the nested `if`s?

Reply:
I tried changing this to the following. The share and delete options were fine, but this caused the transfer option to break. Transfer ownership was only working for private courses. 

I'm not sure how much you want to avoid nested ifs, if it was just supposed to be a trivial change then I suggest we leave it as it is, if it's a problem I'm happy to have another stab at it.

```
handleUserCoursesOnDelete: function(userId, courseOption, done) {
    var pluginManager = pluginmanager.getManager();
    var plugins = pluginManager.getPlugins();
    var CourseContent = require(plugins.content.course.fullPath);
    var courseContent = new CourseContent();
    if (courseOption === null) {
      return done(null);
    }
    courseContent.retrieve({ createdBy: userId }, {}, function (err, results) {
      var currentUser = this.getCurrentUser();
      async.eachSeries(results, function(model, cb) {
        // For a full description of behaviour see https://github.com/adaptlearning/adapt_authoring/pull/2277
        // Transfer ownership of all courses to current user
        courseContent.update({ _id: model._id }, { createdBy: currentUser._id }, function (err, course) {
          if (course._isShared || course._shareWithUsers.length > 0) {
            // Remove current user from the 'shared with' list if applicable
            return courseContent.update({ _id: course._id }, { _shareWithUsers: course._shareWithUsers.pull(currentUser._id) }, cb);
          }
          // Delete private courses if this option was set
          if (courseOption === 'delete') {
            return courseContent.destroy({ _id: course._id }, false, cb);
          }
          // Share private courses with everyone if this option was set, remove current user from the 'shared with' list if applicable
          if (courseOption === 'share') {
            return courseContent.update({ _id: course._id }, { _isShared: true, _shareWithUsers: course._shareWithUsers.pull(currentUser._id) }, cb);
          }
        }.bind(this));
      }.bind(this), done);
    }.bind(this));
  },
```

Reply:
In this case it was less about the nested `if`s and more about confirming in exactly which scenarios the code falls through to line 372. Trying to refactor this sort of highlights the issue.

I'll try and have a stab at this myself.

Reply:
I've got this so far, with the return of Thomas' switch statement:
```js
const currentUserId = this.getCurrentUser()._id;
courseContent.retrieve({ createdBy: userId }, {}, (err, results) => {
  async.eachSeries(results, (model, cb) => {
    // For a full description of behaviour see https://github.com/adaptlearning/adapt_authoring/pull/2277
    switch (courseOption) {
      case 'share':
        // Share private courses with everyone if this option was set, remove the current user from the 'shared with' list if applicable
        courseContent.update({ _id: model._id }, { createdBy: currentUserId, _isShared: true, _shareWithUsers: model._shareWithUsers.pull(currentUserId) }, cb);
        break;
      case 'delete':
        // Delete private courses if this option was set
        if (!model._isShared && !model._shareWithUsers.length) {
          courseContent.destroy({ _id: model._id }, false, cb);
          break;
        }
      case 'transfer':
        // Remove the current user from the 'shared with' list if applicable
        courseContent.update({ _id: model._id }, { createdBy: currentUserId, _shareWithUsers: model._shareWithUsers.pull(currentUserId) }, cb);
    }
  }, done);
});
```
💀 **NOTE**: whichever approach we go for, the code needs updating to to handle an undefined `_shareWithUsers`; this property isn't present in the existing versions of the framework and currently errors.

Reply:
How's about this one then? Has an early return, handles frameworks without `_shareWithUsers` and I've tried to preserve your comments:
```js
const currentUserId = this.getCurrentUser()._id;
courseContent.retrieve({ createdBy: userId }, {}, (err, results) => {
  async.eachSeries(results, (model, cb) => {
    // For a full description of behaviour see https://github.com/adaptlearning/adapt_authoring/pull/2277
    const sharedWithList = model._shareWithUsers;
    const isPartiallyShared = sharedWithList && sharedWithList.length;
    // Delete private courses if this option was set
    if (courseOption === 'delete' && !model._isShared && !isPartiallyShared) {
      courseContent.destroy({ _id: model._id }, false, cb);
      return;
    }
    // Transfer ownership of all courses to current user
    let update = { createdBy: currentUserId };
    // Remove the current user from the 'shared with' list if applicable
    if (isPartiallyShared) {
      update._shareWithUsers = sharedWithList.pull(currentUserId);
    }
    // Share private courses with everyone if this option was set
    if (courseOption === 'share') {
      update._isShared = true;
    }
    courseContent.update({ _id: model._id }, update, cb);
  }, done);
});
```

Reply:
That is looking really good to me (can't test right now though). Seems much more elegant.